lib/Parse/ParsePattern.cpp

```diff
@@ -786,15 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
   auto result = parsePattern();
   
   // Now parse an optional type annotation.
-  if (consumeIf(tok::colon)) {
+  if (Tok.is(tok::colon)) {
+    SourceLoc colonLoc = consumeToken(tok::colon);
+    
     if (result.isNull())  // Recover by creating AnyPattern.
       result = makeParserErrorResult(new (Context) AnyPattern(PreviousLoc));
     
     ParserResult<TypeRepr> Ty = parseType();
     if (Ty.hasCodeCompletion())
       return makeParserCodeCompletionResult<Pattern>();
-    if (Ty.isNull())
+    if (!Ty.isNull()) {
+      // Attempt to diagnose initializer calls incorrectly written
+      // as typed patterns, such as "var x: [Int]()".
+      if (Tok.isFollowingLParen()) {
+        SourceLoc lParenLoc, rParenLoc;
+        SmallVector<Expr *, 2> args;
+        SmallVector<Identifier, 2> argLabels;
+        SmallVector<SourceLoc, 2> argLabelLocs;
+        Expr *trailingClosure;
+        ParserStatus status = parseExprList(tok::l_paren, tok::r_paren,
+                                            /*isPostfix=*/true,
+                                            /*isExprBasic=*/false,
+                                            lParenLoc, args, argLabels,
+                                            argLabelLocs, rParenLoc,
+                                            trailingClosure);
+        if (status.isSuccess()) {
+          // Suggest replacing : with = (with proper whitespace).
+          llvm::StringRef replacement = " =";
+          if (SourceMgr.extractText({colonLoc.getAdvancedLoc(-1), 1}) == " ") {
+            replacement = "=";
+          }
+          diagnose(lParenLoc, diag::initializer_as_typed_pattern)
+            .highlight({Ty.get()->getStartLoc(), rParenLoc})
+            .fixItReplace(colonLoc, replacement);
+          result.setIsParseError();
```

Comment:
Hm, I guess there is a question of whether we want PatternBindingDecl parsing to generally recognize incorrect `var x: «expr»` and treat it as `var x = «expr»` instead.

(The fix as written is pretty simple because it only handles exprs that look like initializer calls. It handles `var x: [Int]()`, but it won't handle `var x: Unmanaged.fromOpaque(y)` or `var x: MemoryLayout<Int>.stride`, for instance.)

In general, if the user made a typo and wrote `:` instead of `=`, the parser might only notice a problem after *successfully* parsing a typed pattern binding. So, unless we want to try parsing `var x = «expr»` *first* every time we look at a pattern (even if we see a `:`), I think this would need to be hoisted even further out than the caller of `parseTypedPattern()`, because it needs to invoke all the logic in `parseExpr`/`parseExprPostfix`, and it might need to do so even if parseTypedPattern succeeded.

Which is to say: the fix as written is pretty simple and self-contained. After reading some more of the parser implementation, I'm inclined to think producing better results than this would be considerably more involved. Feel free to prove me wrong, if you can think of a clean way to do it!